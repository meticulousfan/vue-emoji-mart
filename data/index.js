var buildSearch = require('../src/utils/build-search')
var data = require('./data')

function uncompress (list) {
  for (var short_name in list) {
    var datum = list[short_name]

    if (!datum.short_names) datum.short_names = []
    datum.short_names.unshift(short_name)

    datum.sheet_x = datum.sheet[0]
    datum.sheet_y = datum.sheet[1]
    delete datum.sheet

    if (!datum.text) datum.text = ''
    if (datum.added_in !== null && !datum.added_in) datum.added_in = '6.0'

    datum.search = buildSearch({
      short_names: datum.short_names,
      name: datum.name,
      keywords: datum.keywords,
      emoticons: datum.emoticons
    })

    datum.search = datum.search.join(',')
  }
}

uncompress(data.emojis)
uncompress(data.skins)

module.exports = data
