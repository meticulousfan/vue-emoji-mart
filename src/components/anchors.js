import React from 'react'
import PropTypes from 'prop-types'

import SVGs from '../svgs'

export default class Anchors extends React.PureComponent {
  constructor(props) {
    super(props)

    const { categories } = props

    const defaultCategory = categories.filter(category => category.first)[0]

    this.state = {
      selected: defaultCategory.name,
    }

    this.handleClick = this.handleClick.bind(this)
  }

  handleClick(e) {
    var index = e.currentTarget.getAttribute('data-index')
    var { categories, onAnchorClick } = this.props

    onAnchorClick(categories[index], index)
  }

  render() {
    var { categories, onAnchorClick, color, i18n } = this.props,
      { selected } = this.state

    return (
      <div className="emoji-mart-anchors">
        {categories.map((category, i) => {
          var { name, anchor } = category,
            isSelected = name == selected

          if (anchor === false) {
            return null
          }

          return (
            <span
              key={name}
              title={i18n.categories[name.toLowerCase()]}
              data-index={i}
              onClick={this.handleClick}
              className={`emoji-mart-anchor ${isSelected
                ? 'emoji-mart-anchor-selected'
                : ''}`}
              style={{ color: isSelected ? color : null }}
            >
              <div dangerouslySetInnerHTML={{ __html: SVGs[name] }} />
              <span
                className="emoji-mart-anchor-bar"
                style={{ backgroundColor: color }}
              />
            </span>
          )
        })}
      </div>
    )
  }
}

Anchors.propTypes = {
  categories: PropTypes.array,
  onAnchorClick: PropTypes.func,
}

Anchors.defaultProps = {
  categories: [],
  onAnchorClick: () => {},
}
