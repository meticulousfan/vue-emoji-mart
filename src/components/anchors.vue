<template>

<div class="emoji-mart-anchors">
  <span
    v-for="category in categories"
    :key="category.id"
    :class="{ 'emoji-mart-anchor': true, 'emoji-mart-anchor-selected': category.id == activeCategory.id }"
    :style="{ 'color': (category.id == activeCategory.id ? color : '') }"
    :title="i18n.categories[category.id]"
    @click="$emit('click', category)">
    <div v-html="svgs[category.id]"/>
    <span class="emoji-mart-anchor-bar" :style="{ backgroundColor: color }"></span>
  </span>
</div>

</template>

<script>

import svgs from '../svgs'

export default {
  props: {
    i18n: {
      type: Object,
      required: true
    },
    color: {
      type: String
    },
    categories: {
      type: Array,
      required: true
    },
    activeCategory: {
      type: Object,
      default() {
        return {}
      }
    }
  },
  created() {
    this.svgs = svgs
  }
}

</script>

<style scoped>

.emoji-mart-anchors {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  padding: 0 6px;
  color: #858585;
  line-height: 0;
}

.emoji-mart-anchor {
  position: relative;
  display: block;
  flex: 1 1 auto;
  text-align: center;
  padding: 12px 4px;
  overflow: hidden;
  transition: color .1s ease-out;
}
.emoji-mart-anchor:hover,
.emoji-mart-anchor-selected {
  color: #464646;
}

.emoji-mart-anchor-selected .emoji-mart-anchor-bar {
  bottom: 0;
}

.emoji-mart-anchor-bar {
  position: absolute;
  bottom: -3px; left: 0;
  width: 100%; height: 3px;
  background-color: #464646;
}

.emoji-mart-anchors i {
  display: inline-block;
  width: 100%;
  max-width: 22px;
}

</style>

<style>

.emoji-mart-anchors svg {
  fill: currentColor;
  max-height: 18px;
}

</style>
