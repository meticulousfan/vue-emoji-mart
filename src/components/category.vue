<template>

<div :class="{ 'emoji-mart-category': true, 'emoji-mart-no-results': !hasResults }" v-if="isVisible && (isSearch || hasResults)">
  <div class="emoji-mart-category-label">
    <span>{{ i18n.categories[id] }}</span>
  </div>

  <nimble-emoji
    v-for="emoji in emojis"
    :key="emoji.id || emoji"
    :data="data"
    :emoji="emoji"
    :native="emojiProps.native"
    :skin="emojiProps.skin"
    :set="emojiProps.set"
    :size="emojiProps.size"
    :sheet-size="emojiProps.sheetSize"
    :force-size="emojiProps.forceSize"
    :tooltip="emojiProps.tooltip"
    :background-image-fn="emojiProps.backgroundImageFn"
    @click="emojiProps.onClick"
    @mouseenter="emojiProps.onEnter"
    @mouseleave="emojiProps.onLeave"
  />

  <div v-if="!hasResults">
    <nimble-emoji
      :data="data"
      :size="emojiProps.size"
      emoji="sleuth_or_spy"
      :native="emojiProps.native"
      :skin="emojiProps.skin"
      :set="emojiProps.set"
      :sheet-size="emojiProps.sheetSize"
      :background-image-fn="emojiProps.backgroundImageFn"
    />
    <div class="emoji-mart-no-results-label">{{ i18n.notfound }}</div>
  </div>
</div>

</template>

<script>

import NimbleEmoji from './emoji/nimbleEmoji'

export default {
  props: {
    data: {
      type: Object,
      required: true
    },
    i18n: {
      type: Object,
      required: true
    },
    id: {
      type: String,
      required: true
    },
    name: {
      type: String,
      required: true
    },
    emojis: {
      type: Array
    },
    emojiProps: {
      type: Object,
      required: true
    }
  },
  computed: {
    isVisible() {
      return !!this.emojis
    },
    isSearch() {
      return this.name == 'Search'
    },
    hasResults() {
      return this.emojis.length > 0
    }
  },
  components: {
    NimbleEmoji
  }
}

</script>

<style scoped>

.emoji-mart-category {
  position: relative;
}

.emoji-mart-category-label {
  position: sticky;
  top: 0;
}

.emoji-mart-category .emoji-mart-emoji:before {
  z-index: 0;
  content: "";
  position: absolute;
  top: 0; left: 0;
  width: 100%; height: 100%;
  background-color: #f4f4f4;
  border-radius: 100%;
  opacity: 0;
}

.emoji-mart-category .emoji-mart-emoji:hover:before {
  opacity: 1;
}

.emoji-mart-category-label {
  z-index: 2;
  position: relative;
  position: -webkit-sticky;
  position: sticky;
  top: 0;
}

.emoji-mart-category-label span {
  display: block;
  width: 100%;
  font-weight: 500;
  padding: 5px 6px;
  background-color: #fff;
  background-color: rgba(255, 255, 255, .95);
}

.emoji-mart-no-results {
  font-size: 14px;
  text-align: center;
  padding-top: 70px;
  color: #858585;
}

.emoji-mart-no-results .emoji-mart-category-label {
  display: none;
}

.emoji-mart-no-results .emoji-mart-no-results-label {
  margin-top: .2em;
}

.emoji-mart-no-results .emoji-mart-emoji:hover:before {
  content: none;
}

</style>

<style>

.emoji-mart-category .emoji-mart-emoji span {
  z-index: 1;
  position: relative;
  text-align: center;
  cursor: default;
}

</style>
