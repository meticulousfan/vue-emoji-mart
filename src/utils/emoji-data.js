import { getData, sanitize } from './index'

const SHEET_COLUMNS = 52

export class EmojiData {

  constructor(emoji, skin, set, data) {
    this._key = emoji
    this._data = Object.assign({}, getData(
        emoji, skin, set, data))
    this._sanitized = sanitize(this._data)
    for (let key in this._sanitized) {
      this[key] = this._sanitized[key]
    }
    Object.freeze(this)
  }

  getPosition() {
      let multiply = 100 / (SHEET_COLUMNS - 1),
          x = multiply * this._data.sheet_x,
          y = multiply * this._data.sheet_y
      return `${x}% ${y}%`
  }

}

export class EmojiView {

  /**
   * emoji - EmojiData to display
   * set - string, emoji set name
   * native - boolean, whether to render native emoji
   * fallback - fallback function to render missing emoji, optional
   */
  constructor(emoji, set, native, fallback) {
    this._emoji = emoji
    this._native = native
    this._set = set
    this._fallback = fallback

    this.canRender = this._canRender()
    this.cssClass = this._cssClass()
    this.cssStyle = this._cssStyle()
    this.content = this._content()

    Object.freeze(this)
  }

  _canRender() {
    return this._isCustom() || this._isNative() || this._hasEmoji() || this._fallback
  }

  _cssClass() {
    return [
      'emoji-set-' + this._set,
      'emoji-type-' + this._emojiType()
    ]
  }

  _cssStyle() {
    if (this._isCustom()) {
      return {
        backgroundImage: 'url(' + this._emoji._data.imageUrl + ')',
        backgroundSize: '100%',
      }
    }
    if (this._hasEmoji()) {
      return {
        backgroundPosition: this._emoji.getPosition()
      }
    }
    return {}
  }

  _content() {
    if (this._isCustom()) {
      return ''
    }
    if (this._isNative()) {
      return this._emoji.native
    }
    if (this._hasEmoji()) {
      return ''
    }
    return this._fallback ? this._fallback(this._emoji) : null
  }

  _isNative() {
    return this._native
  }

  _isCustom() {
    return this._emoji.custom
  }

  _hasEmoji() {
    return this._emoji._data && this._emoji._data['has_img_' + this._set]
  }

  _emojiType() {
    if (this._isCustom()) {
      return 'custom';
    }
    if (this._isNative()) {
      return 'native';
    }
    if (this._hasEmoji()) {
      return 'image';
    }
    return 'fallback';
  }

}
