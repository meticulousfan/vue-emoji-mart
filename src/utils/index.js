import { buildSearch } from './data'
import stringFromCodePoint from '../polyfills/stringFromCodePoint'

const _JSON = JSON

const COLONS_REGEX = /^(?:\:([^\:]+)\:)(?:\:skin-tone-(\d)\:)?$/
const SKINS = ['1F3FA', '1F3FB', '1F3FC', '1F3FD', '1F3FE', '1F3FF']

function unifiedToNative(unified) {
  var unicodes = unified.split('-'),
    codePoints = unicodes.map((u) => `0x${u}`)

  return stringFromCodePoint.apply(null, codePoints)
}

function sanitize(emoji) {
  var {
      name,
      short_names,
      skin_tone,
      skin_variations,
      emoticons,
      unified,
      custom,
      imageUrl,
    } = emoji,
    id = emoji.id || short_names[0],
    colons = `:${id}:`

  if (custom) {
    return {
      id,
      name,
      colons,
      emoticons,
      custom,
      imageUrl,
    }
  }

  if (skin_tone) {
    colons += `:skin-tone-${skin_tone}:`
  }

  return {
    id,
    name,
    colons,
    emoticons,
    unified: unified.toLowerCase(),
    skin: skin_tone || (skin_variations ? 1 : null),
    native: unifiedToNative(unified),
  }
}

function cloneEmoji(emoji) {
  if (typeof emoji === 'string') {
    return emoji;
  }

  return Object.assign({}, emoji);
}

function uniq(arr) {
  return arr.reduce((acc, item) => {
    if (acc.indexOf(item) === -1) {
      acc.push(item)
    }
    return acc
  }, [])
}

function intersect(a, b) {
  const uniqA = uniq(a)
  const uniqB = uniq(b)

  return uniqA.filter((item) => uniqB.indexOf(item) >= 0)
}

function deepMerge(a, b) {
  var o = {}

  for (let key in a) {
    let originalValue = a[key],
      value = originalValue

    if (b.hasOwnProperty(key)) {
      value = b[key]
    }

    if (typeof value === 'object') {
      value = deepMerge(originalValue, value)
    }

    o[key] = value
  }

  return o
}

// https://github.com/sonicdoe/measure-scrollbar
function measureScrollbar() {
  if (typeof document == 'undefined') return 0
  const div = document.createElement('div')

  div.style.width = '100px'
  div.style.height = '100px'
  div.style.overflow = 'scroll'
  div.style.position = 'absolute'
  div.style.top = '-9999px'

  document.body.appendChild(div)
  const scrollbarWidth = div.offsetWidth - div.clientWidth
  document.body.removeChild(div)

  return scrollbarWidth
}

export {
  sanitize,
  uniq,
  intersect,
  deepMerge,
  unifiedToNative,
  measureScrollbar,
}
