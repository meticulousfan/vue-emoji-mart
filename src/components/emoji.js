import React from 'react'
import PropTypes from 'prop-types'
import data from '../../data'

import { getData, getSanitizedData, unifiedToNative } from '../utils'

const SHEET_COLUMNS = 49

const _getPosition = (props) => {
  var { sheet_x, sheet_y } = _getData(props),
      multiply = 100 / (SHEET_COLUMNS - 1)

  return `${multiply * sheet_x}% ${multiply * sheet_y}%`
}

const _getData = (props) => {
  var { emoji, skin, set } = props
  return getData(emoji, skin, set)
}

const _getSanitizedData = (props) => {
  var { emoji, skin, set } = props
  return getSanitizedData(emoji, skin, set)
}

export default class Emoji extends React.PureComponent {

  constructor (props) {
    super(props)

    this.handleClick = this.handleClick.bind(this)
    this.handleEnter = this.handleEnter.bind(this)
    this.handleLeave = this.handleLeave.bind(this)
  }

  handleClick (e) {
    if (!this.props.onClick) { return }

    const { onClick } = this.props,
        emoji = _getSanitizedData(this.props)

    onClick(emoji, e)
  }

  handleEnter (e) {
    if (!this.props.onOver) { return }

    const { onOver } = this.props,
        emoji = _getSanitizedData(this.props)

    onOver(emoji, e)
  }

  handleLeave (e) {
    if (!this.props.onLeave) { return }

    const { onLeave } = this.props,
        emoji = _getSanitizedData(this.props)

    onLeave(emoji, e)
  }

  render () {
    var props = this.props;

    var { unified, custom, imageUrl } = _getData(props),
        style = {},
        children = props.children

    if (!unified && !custom) {
      return null
    }

    if (props.native && unified) {
      style = { fontSize: props.size }
      children = unifiedToNative(unified)

      if (props.forceSize) {
        style.display = 'inline-block'
        style.width = props.size
        style.height = props.size
      }
    } else if (custom) {
      style = {
        width: props.size,
        height: props.size,
        display: 'inline-block',
        backgroundImage: `url(${imageUrl})`,
        backgroundSize: '100%',
      }
    } else {
      let setHasEmoji = _getData(props)[`has_img_${props.set}`]

      if (!setHasEmoji) {
        return null
      }

      style = {
        width: props.size,
        height: props.size,
        display: 'inline-block',
        backgroundImage: `url(${props.backgroundImageFn(props.set, props.sheetSize)})`,
        backgroundSize: `${100 * SHEET_COLUMNS}%`,
        backgroundPosition: _getPosition(props),
      }
    }

    return <span
      key={props.emoji.id || props.emoji}
      onClick={this.handleClick}
      onMouseEnter={this.handleEnter}
      onMouseLeave={this.handleLeave}
      className='emoji-mart-emoji'>
      <span style={style}>{children}</span>
    </span>
  }

}

Emoji.propTypes = {
  onOver: PropTypes.func,
  onLeave: PropTypes.func,
  onClick: PropTypes.func,
  backgroundImageFn: PropTypes.func,
  native: PropTypes.bool,
  forceSize: PropTypes.bool,
  skin: PropTypes.oneOf([1, 2, 3, 4, 5, 6]),
  sheetSize: PropTypes.oneOf([16, 20, 32, 64]),
  set: PropTypes.oneOf(['apple', 'google', 'twitter', 'emojione', 'messenger', 'facebook']),
  size: PropTypes.number.isRequired,
  emoji: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
  ]).isRequired,
}

Emoji.defaultProps = {
  skin: 1,
  set: 'apple',
  sheetSize: 64,
  native: false,
  forceSize: false,
  backgroundImageFn: ((set, sheetSize) => `https://unpkg.com/emoji-datasource-${set}@${EMOJI_DATASOURCE_VERSION}/img/${set}/sheets/${sheetSize}.png`),
  onOver: (() => {}),
  onLeave: (() => {}),
  onClick: (() => {}),
}
