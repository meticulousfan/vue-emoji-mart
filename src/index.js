import store from './utils/store'
import { uncompress } from './utils/data'
import frequently from './utils/frequently'

export {
  Picker,
  NimblePicker,
  Emoji,
  NimbleEmoji,
  Category,
  Skins,
} from './components'

export { EmojiIndex } from './utils/emoji-data'
export { uncompress, store, frequently }
