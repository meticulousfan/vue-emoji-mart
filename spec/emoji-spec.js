import { mount } from '@vue/test-utils'

import data from '../data/all.json'
import { EmojiIndex, EmojiView } from '../src/utils/emoji-data'
import { Emoji, NimbleEmoji } from '../src/components'

describe('Emoji', () => {
  const index = new EmojiIndex(data)
  const emoji = mount(Emoji, {
    propsData: {
      emoji: 'point_up',
    },
  })

  it('works', () => {
    expect(emoji.isVueInstance()).toBeTruthy()
    expect(emoji.html()).toContain('emoji-set-apple emoji-type-image')
    expect(emoji.vm.emojiObject.id).toBe('point_up')
  })

  test('renders correctly', () => {
    expect(emoji.element).toMatchSnapshot()
  })

  test('renders correctly native emoji', () => {
    emoji.setProps({ native: true })
    expect(emoji.element).toMatchSnapshot()
  })

  it('has correct emoji and emoji view', () => {
    let expectedEmoji = index.emoji('point_up')
    let expectedView = new EmojiView(expectedEmoji, 1, 'apple', true, undefined)
    expect(emoji.vm.emojiObject).toEqual(expectedEmoji)
    expect(emoji.vm.view).toEqual(expectedView)
  })

  it('has correct sanitized data', () => {
    expect(emoji.vm.sanitizedData).toEqual({
      colons: ':point_up:',
      emoticons: undefined,
      id: 'point_up',
      name: 'White Up Pointing Index',
      native: '☝️',
      skin: 1,
      unified: '261d-fe0f',
    })
  })
})

describe('Emoji native', () => {
  const index = new EmojiIndex(data)
  const emoji = mount(Emoji, {
    propsData: {
      native: true,
      emoji: 'point_up',
    },
  })

  test('renders correctly native emoji', () => {
    expect(emoji.element).toMatchSnapshot()
  })

  it('has correct emoji and emoji view', () => {
    let expectedEmoji = index.emoji('point_up')
    let expectedView = new EmojiView(expectedEmoji, 1, 'apple', true, undefined)
    expect(emoji.vm.emojiObject).toEqual(expectedEmoji)
    expect(emoji.vm.view).toEqual(expectedView)
  })
})

describe('Emoji title', () => {
  const emoji = mount(Emoji, {
    propsData: {
      emoji: 'point_up',
    },
  })

  it('has no title by default', () => {
    expect(emoji.vm.title).toBe(null)
  })

  it('has short name as a title if tooltip property is set to true', () => {
    emoji.setProps({ tooltip: true })
    expect(emoji.vm.title).toBe('point_up')
  })

  it('does not have title if tooltop property is set to false', () => {
    emoji.setProps({ tooltip: false })
    expect(emoji.vm.title).toBe(null)
  })
})
