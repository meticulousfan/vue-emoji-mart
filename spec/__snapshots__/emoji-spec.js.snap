// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emoji native renders correctly native emoji 1`] = `
<span
  aria-label="☝️, point_up"
  class="emoji-mart-emoji"
>
  <span
    class="emoji-set-apple emoji-type-native"
  >
    ☝️
  </span>
</span>
`;

exports[`Emoji renders correctly 1`] = `
<span
  aria-label="☝️, point_up"
  class="emoji-mart-emoji"
>
  <span
    class="emoji-set-apple emoji-type-image"
    style="background-position: 94.74% 12.28%;"
  >
    
  </span>
</span>
`;

exports[`Emoji renders correctly native emoji 1`] = `
<span
  aria-label="☝️, point_up"
  class="emoji-mart-emoji"
>
  <span
    class="emoji-set-apple emoji-type-native"
    style="background-position: 94.74% 12.28%;"
  >
    ☝️
  </span>
</span>
`;
